external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/KafkaDataConsumer.scala

```diff
@@ -0,0 +1,381 @@
+
+}
+
+private[kafka010]
+object KafkaDataConsumer extends Logging {
+
+  private case class CachedKafkaDataConsumer[K, V](internalConsumer: InternalKafkaConsumer[K, V])
+    extends KafkaDataConsumer[K, V] {
+    assert(internalConsumer.inUse) // make sure this has been set to true
+    override def release(): Unit = { KafkaDataConsumer.release(internalConsumer) }
+  }
+
+  private case class NonCachedKafkaDataConsumer[K, V](internalConsumer: InternalKafkaConsumer[K, V])
+    extends KafkaDataConsumer[K, V] {
+    override def release(): Unit = { internalConsumer.close() }
+  }
+
+  private case class CacheKey(groupId: String, topicPartition: TopicPartition)
+
+  // Don't want to depend on guava, don't want a cleanup thread, use a simple LinkedHashMap
+  private var cache: ju.Map[CacheKey, ju.List[InternalKafkaConsumer[_, _]]] = null
+
+  /**
+   * Must be called before acquire, once per JVM, to configure the cache.
+   * Further calls are ignored.
+   * */
+  def init(
+      initialCapacity: Int,
+      maxCapacity: Int,
+      loadFactor: Float): Unit = synchronized {
+    if (null == cache) {
+      logInfo(s"Initializing cache $initialCapacity $maxCapacity $loadFactor")
+      cache = new ju.LinkedHashMap[CacheKey, ju.List[InternalKafkaConsumer[_, _]]](
+        initialCapacity, loadFactor, true) {
+        override def removeEldestEntry(
+          entry: ju.Map.Entry[CacheKey, ju.List[InternalKafkaConsumer[_, _]]]): Boolean = {
+          if (this.size > maxCapacity) {
+            try {
+              entry.getValue.asScala.foreach { _.close() }
+            } catch {
+              case x: KafkaException =>
+                logError("Error closing oldest Kafka consumer", x)
+            }
+            true
+          } else {
+            false
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Get a cached consumer for groupId, assigned to topic and partition.
+   * If matching consumer doesn't already exist, will be created using kafkaParams.
+   * The returned consumer must be released explicitly using [[KafkaDataConsumer.release()]].
+   *
+   * Note: This method guarantees that the consumer returned is not currently in use by anyone
+   * else. Within this guarantee, this method will make a best effort attempt to re-use consumers by
+   * caching them and tracking when they are in use.
+   */
+  def acquire[K, V](
+      groupId: String,
+      topicPartition: TopicPartition,
+      kafkaParams: ju.Map[String, Object],
+      context: TaskContext,
+      useCache: Boolean): KafkaDataConsumer[K, V] = synchronized {
+    val key = new CacheKey(groupId, topicPartition)
+    val existingInternalConsumers = Option(cache.get(key))
+      .getOrElse(new ju.LinkedList[InternalKafkaConsumer[_, _]])
+
+    cache.putIfAbsent(key, existingInternalConsumers)
+
+    lazy val newInternalConsumer = new InternalKafkaConsumer[K, V](
+      groupId, topicPartition, kafkaParams)
+
+    if (context != null && context.attemptNumber >= 1) {
```

Comment:
I agree, the comment in the old code was more clear: "just in case the prior attempt failures were cache related"

Reply:
If this is the first attempt then the list can contain elements because other tasks can create other consumers. Imran is right the `why` was written in `KafkaRDD.scala` which is missing here. Additional explanation added.

In short if the possible problematic consumer is not removed from the list then it could infect tasks run later.